plugins/clipboard/plugin.js

```diff
@@ -146,75 +148,96 @@
 			// custom MIME types in clipboard (#4612).
 			// Do it as the first step as the conversion is asynchronous and should hold all further paste processing.
-			var isImagePasteSupported = CKEDITOR.plugins.clipboard.isCustomDataTypesSupported || CKEDITOR.plugins.clipboard.isFileApiSupported;
-			if ( isImagePasteSupported && editor.config.clipboard_handleImages ) {
-				var supportedImageTypes = [ 'image/png', 'image/jpeg', 'image/gif' ],
-					unsupportedTypeMsg = createNotificationMessage( supportedImageTypes ),
-					latestId;
-
-				editor.on( 'paste', function( evt ) {
-					var dataObj = evt.data,
-						data = dataObj.dataValue,
-						dataTransfer = dataObj.dataTransfer;
-
-					// If data empty check for image content inside data transfer. https://dev.ckeditor.com/ticket/16705
-					// Allow both dragging and dropping and pasting images as base64 (#4681).
-					if ( !data && isFileData( evt, dataTransfer ) ) {
-						var file = dataTransfer.getFile( 0 );
-
-						if ( CKEDITOR.tools.indexOf( supportedImageTypes, file.type ) === -1 ) {
-							editor.showNotification( unsupportedTypeMsg, 'info', editor.config.clipboard_notificationDuration );
-
-							return;
-						}
+			var isFilePasteSupported = CKEDITOR.plugins.clipboard.isCustomDataTypesSupported || CKEDITOR.plugins.clipboard.isFileApiSupported,
+				latestId;
+			// File matcher will be used by clipboard base64 handler below (#5095).
+			this.addNotificationFileMatcher( testImageBase64Support );
 
-						var fileReader = new FileReader();
+			editor.on( 'paste', function( evt ) {
+				if ( !isFilePasteSupported ) {
+					return;
+				}
 
-						// Convert image file to img tag with base64 image.
-						fileReader.addEventListener( 'load', function() {
-							evt.data.dataValue = '<img src="' + fileReader.result + '" />';
-							editor.fire( 'paste', evt.data );
-						}, false );
+				var dataObj = evt.data,
+					data = dataObj.dataValue,
+					dataTransfer = dataObj.dataTransfer;
 
-						// Proceed with normal flow if reading file was aborted.
-						fileReader.addEventListener( 'abort', function() {
-							// (#4681)
-							setCustomIEEventAttribute( evt );
-							editor.fire( 'paste', evt.data );
-						}, false );
+				// If data empty check for file inside data transfer. https://dev.ckeditor.com/ticket/16705
+				if ( data || !isFileData( evt, dataTransfer ) ) {
+					return;
+				}
 
-						// Proceed with normal flow if reading file failed.
-						fileReader.addEventListener( 'error', function() {
-							// (#4681)
-							setCustomIEEventAttribute( evt );
-							editor.fire( 'paste', evt.data );
-						}, false );
+				var file = dataTransfer.getFile( 0 );
 
-						fileReader.readAsDataURL( file );
+				// (#5095)
+				if ( !isFileTypeSupported( file ) ) {
+					displayUnsupportedFileTypeNotification( file.type );
+					return;
+				}
 
-						latestId = dataObj.dataTransfer.id;
+				// Convert image file (if present) to base64 string.
```

Comment:
I'm wondering if it would not be more readable and semantically correct to split these two things (showing notification and the whole image pasting):

```javascript
editor.on( 'paste', function( evt ) {
	// Notification showing logic.
}, null, null, 1 );

editor.on( 'paste', function( evt ) {
	// Image handling logic.
}, null, null, 1 );
```